### ISSUE TITLE
No support for lazily loaded submodules

### ISSUE THREAD

**Issue by thomasaarholt** [microsoft/pyright#3595]
**Describe the bug**
[PEP562](https://peps.python.org/pep-0562/) allows for lazy loading of submodules. See the second code block under [Rationale](https://peps.python.org/pep-0562/#rationale). 

We implemented this in hyperspy last year, and it works great in jupyter lab. However, in VSCode, autocompletion of lazily loaded packages fails. 

```python
import hyperspy.api_nogui as hs
hs.sig<TAB> # autocompletes to `hs.signals` in jupyter, not in VSCode
```

On investigation I noticed that in pyright/pylance raises a warning on many of the the entries in `__all__` in [`hyperspy.api_nogui`](https://github.com/hyperspy/hyperspy/blob/c868840b6aeab28f43f961ff50d93ff10beac3f9/hyperspy/api_nogui.py#L108):
```python
__all__ = [
    'datasets',
    'eds',
    ...
    'transpose',
    '__version__',
    ]
```
 `"datasets" is specified in __all__ but is not present in modulePylancereportUnsupportedDunderAll`

I then checked the example given in PEP562 (second code block), which I have made into a tiny package and zipped: [pep562.zip](https://github.com/microsoft/pyright/files/8949134/pep562.zip)

If you inspect `__init__.py`, you will see that `__all__` raises the same warning. The other warning I believe is because pyright doesn't (yet?) support python operations on `__all__`, meaning that we should explicitly type them out even if they are importable from elsewhere. Is that right?

<img width="999" alt="image" src="https://user-images.githubusercontent.com/2721423/174807181-a6832171-2e9f-4809-aa5f-fd70886d5c24.png">

I noticed issue #3304, and while I believe the concerns raised there are valid (the code referenced in the `email` and `xml` packages is older than PEP562), I believe that this is a separate issue.

**To Reproduce**
See code example above

**Expected behavior**
Pyright either supporting lazy loading or raising a relevant warning.

**VS Code extension or command-line**
VSCode Python extension pre-release v2022.9.11681004

**Comment by erictraut** [comment]
These behaviors (both lazy loading of modules and dynamic manipulation of `__all__`) are too dynamic for a static type checker. We don't have any plans to support these behaviors.

I presume that the desire to load modules lazily is for performance reasons? If so, you can use an `if TYPE_CHECKING:` block to provide a static implementation with the dynamic (lazy) implementation in the `else:` block.

**Comment by thomasaarholt** [comment]
That's totally fair. You were correct with your performance assumption, and nailed it with your suggestion. That worked great, thank you!
